File: solutions/python/kindergarten-garden/1/kindergarten_garden.py
from itertools import islice 

L_STUDENTS = [
    'Alice',  'Bob',     'Charlie',   'David', 
    'Eve',    'Fred',    'Ginny',     'Harriet', 
    'Ileana', 'Joseph',  'Kincaid',   'Larry'
]

FLOWER_DICT = {
    "C": "Clover",
    "G": "Grass",
    "R": "Radishes",
    "V": "Violets"
}

def batched(iterable, n, *, strict=False):
    # batched('ABCDEFG', 3) → ABC DEF G
    if n < 1:
        raise ValueError('n must be at least one')
    iterator = iter(iterable)
    while batch := tuple(islice(iterator, n)):
        if strict and len(batch) != n:
            raise ValueError('batched(): incomplete batch')
        yield batch

class Garden:
    def __init__(self, diagram, students=None):
        self.students = sorted(students) if students is not None else L_STUDENTS
        g = [ ["".join(b) for b in batched(it, 2)] for it in diagram.split("\n")]
        self.garden = [ "".join(item) for item in zip(*g)]

    def plants(self, student):
        pl = self.garden[self.students.index(student)]
        return [FLOWER_DICT[item] for item in pl]

File: solutions/java/kindergarten-garden/1/src/main/java/KindergartenGarden.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

class KindergartenGarden {

    public final static String [] students;
    static {
        students = new String[] {
            "Alice", "Bob", "Charlie", "David", "Eve", "Fred",
            "Ginny", "Harriet", "Ileana", "Joseph", "Kincaid", "Larry"
        };
    }

    private HashMap<String, List<Plant>> mapGarden;

    KindergartenGarden(String garden) {

        String[] gardenRows = garden.split("\n");
        if (gardenRows.length != 2) throw new IllegalArgumentException("Input must be separated by \\n");
        if (gardenRows[0].length() != gardenRows[1].length()) throw new IllegalArgumentException("Input must be two equal-length");

        this.mapGarden = new HashMap<String, List<Plant>>();
        for (int i = 0, n = gardenRows[1].length(); i < n; i+=2) {
            List<Plant> group = new ArrayList<Plant>(4);
            for (int j = 0; j < 2; ++j) {
                group.add(Plant.getPlant(gardenRows[j].charAt(i)));
                group.add(Plant.getPlant(gardenRows[j].charAt(i + 1)));
            }
            this.mapGarden.put(students[i/2], group);
        }
    }

    List<Plant> getPlantsOfStudent(String student) {
        return this.mapGarden.get(student);
    }

}
